doc/high_level_design.md

```diff
@@ -0,0 +1,13 @@
+## Miniscript High level design
+
+![alt text](./miniscript_high_level_arch.png)
+
+As of 10.0.0 rust-miniscript is roughly divided into following components:
+
+* **Miniscript Core**(<span style="color:green">Green box</span>): This is the core of the library. It contains the core data structures and algorithms for miniscript and output descriptors.
+All the users are expected to interact with the library using the descriptor APIs. Raw miniscript should be used only if you know what you are doing.
+* **Policy**(<span style="color:red">Red box</span>): This is the policy language for miniscript.
+    - It is a simple language which allows users to specify the policy for a descriptor. This simple user provided policy is referred as concrete policy. The miniscript compiler compiles this concrete policy into a descriptor.
+    - The policy also provides a semantic policy for analysis of descriptors and miniscript. Users can use the APIs in the module to analyze the descriptors and miniscript for various properties.
```

Comment:
In 78c9df6ffadc6f45dfeb0dca571e71652934d7de:

I think we should say

```
- The policy module also provides a "semantic policy" for analysis of descriptors and miniscript. Unlike the concrete policy, which is implemented and used by other projects, the semantic policy is purely part of rust-miniscript. The APIs on this object can be used to analyze descriptors and miniscripts for various properties.
```

The main thing is to emphasize that there are actually two policy languages here; one is "the miniscript policy language" and the other is an analysis tool.

Arguably we should move the semantic policy out of the policy module ... and maybe rename it to `Abstract` which is what I originally wanted to do, but avoided because `abstract` (lowercase) is a reserved Rust keyword ... and maybe drop the FromStr  and serde impls for it because you shouldn't ever be serializing a semantic policy.

Reply:
Agreed, we move things so that semantic policy and concrete policy are not confused.

Reply:
Updated.